src/main.cpp

```diff
@@ -5810,14 +5788,45 @@ bool SendMessages(CNode* pto)
             }
             pto->vInventoryBlockToSend.clear();
 
-            // Determine transactions to relay
-            bool fSendTrickle = pto->fWhitelisted;
-            if (pto->nNextInvSend < nNow) {
-                fSendTrickle = true;
-                // Use half the delay for outbound peers, as there is less privacy concern for them.
-                pto->nNextInvSend = PoissonNextSend(nNow, INVENTORY_BROADCAST_INTERVAL >> !pto->fInbound);
+            // Respond to BIP35 mempool requests
+            if (fSendTrickle && pto->fSendMempool) {
+                std::vector<uint256> vtxid;
+                mempool.queryHashes(vtxid);
+                pto->fSendMempool = false;
+                CAmount filterrate = 0;
+                {
+                    LOCK(pto->cs_feeFilter);
+                    filterrate = pto->minFeeFilter;
+                }
+
+                LOCK(pto->cs_filter);
+
+                BOOST_FOREACH(const uint256& hash, vtxid) {
+                    CInv inv(MSG_TX, hash);
+                    if (pto->vInventoryTxToSend.count(hash)) continue;
+                    if (filterrate) {
+                        CFeeRate feeRate;
+                        mempool.lookupFeeRate(hash, feeRate);
+                        if (feeRate.GetFeePerK() < filterrate)
+                            continue;
+                    }
+                    if (pto->pfilter) {
+                        CTransaction tx;
+                        bool fInMemPool = mempool.lookup(hash, tx);
+                        if (!fInMemPool) continue; // another thread removed since queryHashes, maybe...
+                        if (!pto->pfilter->IsRelevantAndUpdate(tx)) continue;
+                    }
+                    pto->filterInventoryKnown.insert(hash);
+                    vInv.push_back(inv);
+                    if (vInv.size() == MAX_INV_SZ) {
+                        pto->PushMessage(NetMsgType::INV, vInv);
+                        vInv.clear();
+                    }
+                }
             }
-            if (fSendTrickle) {
+
+            // Determine transactions to relay
+            if (fSendTrickle && vInv.size() < INVENTORY_BROADCAST_MAX) {
```

Comment:
This comparison on `vInv.size()` doesn't quite make sense, as we might have already filled it up and cleared it out above in the `mempool` command handling.  Most likely, if you've responded to a mempool command, you've used up your allotted bandwidth under this logic, but it's not clear to me whether those should really be considered together or not. 

With the change to `mempool` handling happening only when we do our next scheduled send, can we just restore the behavior where we respond with everything in the mempool, and then in those situations bail out on (ever) announcing the transactions that would otherwise have been announced?  Or is that too big a change to make to peers who might be treating responses to the mempool command differently somehow?  (That doesn't exactly make sense to me, as those aren't really distinguishable, but, I dunno...)

My suggestion would also result in a behavior change in the rare situation that a transaction to be relayed has been evicted from the mempool prior to announcement AND the peer has sent a mempool message in the interim, but at first glance that seems like an edge-case optimization to me, rather than an edge-case bug, as you'd probably choose not to relay something if you knew that you'd be evicting it in the very near future.